oxidation/crates/poc/src/parser.rs

```diff
@@ -0,0 +1,336 @@
+        }
+        ty => panic!("{ty:?} encountered"),
+    }
+}
+
+fn quote_serde_as(ty: &Type) -> TokenStream {
+    let serde_as = extract_serde_as(ty)
+        .replace("Vec<u8>", "::serde_with::Bytes")
+        .replace("u8", "_");
+    quote! { #[serde_as(as = #serde_as)] }
+}
+
+enum SerdeAttr {
+    Rename,
+    RenameAll,
+    Tag,
+}
+
+fn quote_serde_attr(attr: SerdeAttr, value: &Option<String>) -> TokenStream {
+    match value {
+        Some(value) => match attr {
+            SerdeAttr::Rename => quote! { #[serde(rename = #value)] },
+            SerdeAttr::RenameAll => quote! { #[serde(rename_all = #value)] },
+            SerdeAttr::Tag => quote! { #[serde(tag = #value)] },
+        },
+        None => quote! {},
+    }
+}
+
+#[derive(Clone, Copy)]
+enum Vis {
+    Public,
+    Private,
+}
+
+#[derive(Deserialize)]
+struct Field {
+    name: String,
+    #[serde(rename = "type")]
+    ty: String,
+    rename: Option<String>,
+}
+
+impl Field {
+    fn quote(&self, vis: Vis) -> TokenStream {
+        let name: Ident = syn::parse_str(&self.name).unwrap_or_else(|_| unreachable!());
+        let ty: Type = syn::parse_str(&self.ty).unwrap_or_else(|e| panic!("{e}"));
+        let rename = quote_serde_attr(SerdeAttr::Rename, &self.rename);
+        let serde_as = quote_serde_as(&ty);
+
+        match vis {
+            Vis::Public => quote! {
+                #rename
+                #serde_as
+                pub #name: #ty
+            },
+            Vis::Private => quote! {
+                #rename
+                #serde_as
+                #name: #ty
+            },
+        }
+    }
+}
+
+fn quote_fields(_fields: &[Field], vis: Vis) -> TokenStream {
+    let mut fields = quote! {};
+
+    for field in _fields {
+        let field = field.quote(vis);
+        fields = quote! {
+            #fields
+            #field,
+        };
+    }
+    fields
+}
+
+#[derive(Deserialize)]
+struct Variant {
+    name: String,
+    rename: Option<String>,
+    fields: Vec<Field>,
+}
+
+impl Variant {
+    fn quote(&self) -> TokenStream {
+        let name: Ident = syn::parse_str(&self.name).unwrap_or_else(|_| unreachable!());
+        let rename = quote_serde_attr(SerdeAttr::Rename, &self.rename);
+        let fields = quote_fields(&self.fields, Vis::Private);
+
+        if self.fields.is_empty() {
+            quote! {
+                #rename
+                #name
+            }
+        } else {
+            quote! {
+                #rename
+                #name {
+                    #fields
+                }
+            }
+        }
+    }
+}
+
+fn quote_variants(_variants: &[Variant]) -> TokenStream {
+    let mut variants = quote! {};
+
+    for variant in _variants {
+        let variant = variant.quote();
+
+        variants = quote! {
+            #variants
+            #variant,
+        };
+    }
+    variants
+}
+
+#[derive(Deserialize)]
+pub(crate) struct Schema {
+    name: String,
+    rename_all: Option<String>,
```

Comment:
as I said, we really want to be explicit in the schema so no `rename`/`rename_all`

Reply:
Done